### ISSUE TITLE
Unable to manually assign grad

### ISSUE THREAD

**Issue by andrei-rusu** [pytorch/pytorch#72515]
### 🐛 Describe the bug

I am trying to implement various RL algorithms using GNNs and Eligibility Traces of the network parameters. I found a couple of examples online, but for some reason I cannot get manual gradient assignment to work.

First thing I tried was something along the lines of:
```python
self.trace = [torch.zeros_like(p.data, requires_grad=False) for p in self.model.parameters()]
...
loss.backward()
for idx, p in enumerate(self.model.parameters()):
    # self.gamma and self.lamda are scalars
    self.trace[idx] = self.gamma * self.lamda * self.trace[idx] + p.grad
    # delta is a scalar
    p.grad = delta * self.trace[idx]
```
The above fails with: '_RuntimeError: assigned grad has data of a different size_'. Upon checking the shapes of the tensors going in the gradient assignment operation, nothing stands out as faulty. Afterwards, I went inside the _tensor.py file and printed out some debug statements in the _grad_ setter, and it seems that, even if my inputted self.trace[idx] has the same shape as p._grad, as should be expected, the _grad_ setter's new_grad argument changes its shape for single-dimensional tensors (e.g. p._grad.shape=(64), self.trace[idx].shape=(64), but new_grad.shape=(1,64)...). Therefore, even the below explicit reshaping fails with the same error (i.e. the reshaping operation executes just fine, as the tensors do actually have the same shape):
```python
...
# delta is a scalar
p.grad = delta * self.trace[idx].reshape(p.grad.shape)
```

It is worth mentioning here that I tried temporarily patching _tensor.py by reshaping new_grad when the above scenario occurs, but this fails later on at the Adam optimizer.step() (when the moments get calculated, a broadcast error occurs).
Second thing I tried was:
```python
eval_gradients = torch.autograd.grad(loss, self.model.parameters())
for idx, p in enumerate(self.model.parameters()):
    self.trace[idx] = self.gamma * self.lamda * self.trace[idx] + eval_gradients[idx]
    p.grad = delta * self.trace[idx]
```
The above fails with '_RuntimeError: assigned grad has data of a different type_'. Upon debugging, it seems that because I am not explicitly calling loss.backward(), p.grad remains None, so of course p.grad will have a different type when compared to self.trace[idx].
I believe the last implementation has worked in the past (reference: https://github.com/karush17/Deep-Eligibility-Traces/blob/master/Pytorch/watkinsq.py), but it obviously fails in my torch-1.9 setup.

Any help with the above will be much appreciated! Thank you very much.

### Versions

PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 15:50:08) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19041-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 850M
Nvidia driver version: 511.65
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] neptune-pytorch-lightning==0.9.7
[pip3] numpy==1.21.1
[pip3] pytorch-lightning==1.4.0
[pip3] torch==1.9.0
[pip3] torch-cluster==1.5.9
[pip3] torch-geometric==1.7.2
[pip3] torch-scatter==2.0.8
[pip3] torch-sparse==0.6.11
[pip3] torch-spline-conv==1.2.1
[pip3] torchaudio==0.9.0
[pip3] torchdiffeq==0.2.2
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.10.0
[conda] blas                      2.112                       mkl    conda-forge
[conda] blas-devel                3.9.0              12_win64_mkl    conda-forge
[conda] cudatoolkit               11.1.1               heb2d755_7    conda-forge
[conda] libblas                   3.9.0              12_win64_mkl    conda-forge
[conda] libcblas                  3.9.0              12_win64_mkl    conda-forge
[conda] liblapack                 3.9.0              12_win64_mkl    conda-forge
[conda] liblapacke                3.9.0              12_win64_mkl    conda-forge
[conda] mkl                       2021.4.0           h0e2418a_729    conda-forge
[conda] mkl-devel                 2021.4.0           h57928b3_730    conda-forge
[conda] mkl-include               2021.4.0           h0e2418a_729    conda-forge
[conda] mypy_extensions           0.4.3            py38haa244fe_4    conda-forge
[conda] neptune-pytorch-lightning 0.9.7                    pypi_0    pypi
[conda] numpy                     1.21.1           py38h09042cb_0    conda-forge
[conda] pytorch                   1.9.0           py3.8_cuda11.1_cudnn8_0    pytorch
[conda] pytorch-cluster           1.5.9           py38_torch_1.9.0_cu111    rusty1s
[conda] pytorch-geometric         1.7.2           py38_torch_1.9.0_cu111    rusty1s
[conda] pytorch-lightning         1.4.0              pyhd8ed1ab_0    conda-forge
[conda] pytorch-scatter           2.0.8           py38_torch_1.9.0_cu111    rusty1s
[conda] pytorch-sparse            0.6.11          py38_torch_1.9.0_cu111    rusty1s
[conda] pytorch-spline-conv       1.2.1           py38_torch_1.9.0_cu111    rusty1s
[conda] torchaudio                0.9.0                      py38    pytorch
[conda] torchdiffeq               0.2.2              pyhd8ed1ab_0    conda-forge
[conda] torchmetrics              0.6.0              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.10.0                   pypi_0    pypi

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by Lezcano** [comment]
Would you mind posting this question in the forum? https://discuss.pytorch.org/
We usually keep the issues for bugs / feature requests and the like, while we use the forum to ask for questions on how to do X in PyTorch.

**Comment by andrei-rusu** [comment]
Well, do you reckon the behavior above is to be expected? I would consider my post a bug report rather than a question...

**Comment by albanD** [comment]
I am not sure to follow what the problem is. If `p.size() == (64)` then the Tensor you give to `p.grad = ` should be of size `(64)` as well.

> the grad setter's new_grad argument changes its shape for single-dimensional tensors

What do you mean by that? The Tensor give is used as-is.

**Comment by andrei-rusu** [comment]
I think I managed to find the issue with the first grad assignment method: The delta scalar was actually a _(1,1)_ Tensor which led to new_grad having the shape  _(1,64)_, despite the fact that self.trace[idx] had the correct (64) shape. The confusing part is that for multidimensional Tensors, the fact that delta is _(1,1)_ does not make a difference, the only problem comes when the data has a one-dimensional shape.
Minimal example to reproduce:
```python
input = torch.Tensor(10, 64)
model = torch.nn.Linear(64,1)
output = model(input).sum()
delta = torch.Tensor(1,1)
trace = [torch.zeros_like(p.data) for p in model.parameters()]
for i, p in enumerate(model.parameters()):
  p.grad = delta * p.grad
```

Anyhow, the above can be fixed by reshaping _delta_ to be a true scalar or taking _delta.item()_, but this behavior seems rather confusing (i.e. even if the shape of the product stays (1,64) for consistency, I would have expected the grad assignment check to know that _(1,64)_ can actually match the _(64)_ data.

The other issue I have, which is why I kept this ticket opened, is that calling _p.grad=_ fails when _p_ does not have its gradient field populated (the type error check). I believe this was permitted a couple of versions ago, and I am not sure why is the call to _backward_ mandatory now?

Thank you for your time! :)

**Comment by andrei-rusu** [comment]
For some reason assigning the grad like that kept complaining on my work setup until a _backward_ was called, but I've just done a _conda update_ and now all seems fixed. Thanks for the help! :)

**Comment by andrei-rusu** [comment]
Ah, I was just able to reproduce my error (partially). The below fails the 'type' check (and actually the 'shape' check too). I thought this was due to the 'grad' attribute being _None_, but it is actually due to the data being _float32_. I can now rest easily, thank you! :)

```python
import torch
model = torch.nn.Linear(64,1)
model.bias.grad = torch.tensor(1, dtype=float)
```
